`AssertionError: offset != 0` when saving tuple of instantiated empty structs with v0.2+

Trying to update Oceananigans.jl to use JLD2.jl v0.2+ but hit this issue. Not sure if this new behavior is expected but thought I'd open an issue to make sure.

With v0.1.14:

```julia
julia> using JLD2

julia> struct Banana end

julia> jldopen("example.jld2", "w") do file
           file["stuff"] = (Banana(), Banana())
       end
(Banana(), Banana())
```

With v0.2.3:

```julia
julia> using JLD2

julia> struct Banana end

julia> jldopen("example.jld2", "w") do file
           file["stuff"] = (Banana(), Banana())
       end
````

```
ERROR: AssertionError: offset != 0
Stacktrace:
 [1] commit_compound(::JLD2.JLDFile{JLD2.MmapIO}, ::Array{Symbol,1}, ::DataType, ::Type{T} where T) at /home/alir/.julia/packages/JLD2/hIQpq/src/data.jl:204
 [2] h5type(::JLD2.JLDFile{JLD2.MmapIO}, ::Type{T} where T, ::Tuple{Banana,Banana}) at /home/alir/.julia/packages/JLD2/hIQpq/src/data.jl:165
 [3] h5type at /home/alir/.julia/packages/JLD2/hIQpq/src/data.jl:170 [inlined]
 [4] write_dataset at /home/alir/.julia/packages/JLD2/hIQpq/src/datasets.jl:528 [inlined]
 [5] write(::JLD2.Group{JLD2.JLDFile{JLD2.MmapIO}}, ::String, ::Tuple{Banana,Banana}, ::JLD2.JLDWriteSession{Dict{UInt64,JLD2.RelOffset}}) at /home/alir/.julia/packages/JLD2/hIQpq/src/groups.jl:119
 [6] write at /home/alir/.julia/packages/JLD2/hIQpq/src/groups.jl:113 [inlined]
 [7] setindex!(::JLD2.Group{JLD2.JLDFile{JLD2.MmapIO}}, ::Tuple{Banana,Banana}, ::String) at /home/alir/.julia/packages/JLD2/hIQpq/src/groups.jl:124
 [8] setindex!(::JLD2.JLDFile{JLD2.MmapIO}, ::Tuple{Banana,Banana}, ::String) at /home/alir/.julia/packages/JLD2/hIQpq/src/JLD2.jl:343
 [9] #11 at ./REPL[13]:2 [inlined]
 [10] jldopen(::var"#11#12", ::String, ::Vararg{String,N} where N; kws::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/alir/.julia/packages/JLD2/hIQpq/src/loadsave.jl:4
 [11] jldopen(::Function, ::String, ::String) at /home/alir/.julia/packages/JLD2/hIQpq/src/loadsave.jl:2
 [12] top-level scope at REPL[13]:1

```

--- a/Project.toml
+++ b/Project.toml
@@ -1,6 +1,6 @@
 name = "JLD2"
 uuid = "033835bb-8acc-5ee8-8aae-3f567f8a3819"
-version = "0.2.3"
+version = "0.2.4"
 
 [deps]
 CodecZlib = "944b1d66-785c-5afd-91f1-9de20f533193"

--- a/src/data.jl
+++ b/src/data.jl
@@ -62,7 +62,7 @@ end
 function hasdata(T::DataType, encounteredtypes=DataType[])
     isempty(T.types) && T.size != 0 && return true
     for ty in T.types
-        hasfielddata(writeas(ty), encounteredtypes) && return true
+        hasfielddata(writeas(ty), copy(encounteredtypes)) && return true
     end
     false
 end

--- a/test/loadsave.jl
+++ b/test/loadsave.jl
@@ -239,3 +239,14 @@ struct UA2{T}; y::T; end
     @test u2 == U2[1, [2.0], 3, ["4"]]
     @test u3 == U3[UA1(1), UA2(2.0), 3, UA1("4")]
 end
+
+
+
+# Test for Issue #247
+@testset "Tuple of Empty Structs" begin
+    fn = joinpath(mktempdir(), "test.jld2")
+    @save fn tup=(EmptyImmutable(), EmptyImmutable())
+    @load fn tup
+
+    @test tup == (EmptyImmutable(), EmptyImmutable())
+end

